index.ts

```diff
@@ -41,22 +51,53 @@ const CUSTOMERS = [
   },
 ];
 
+
+
 // create a function that receives a name of a customer, and returns the balance of that customer.
 // if type of this function is not infered, type it.
-
 // EXTRA: optimize this fucnction to decrease the time complexity.
 // EXTRA 2 : optimize this function to be able to memorize the results of previous calls.
-
 // i: getBalance('John')
 // o: 125
+type Customer = {
+  customer: string;
+  balance: number;
+  id: string;
+}
+
+
+const obtener_saldo = (() => {
+  const cache = new Map<string, number>();
+  
+  return (name: string): string => {
+    if (cache.has(name)) {
+      return cache.get(name)!.toString();
+    }
 
-// consider the following code:
+    const customer = CUSTOMERS.find((c) => c.customer === name);
+    if (customer) {
+      cache.set(name, customer.balance);
+      return customer.balance.toString();
+    } else {
+      return "cliente no encontrado";
+    }
+  };
+})();
 
-const z = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
-const r = Math.ceil(Math.random() * 10);
+console.log(obtener_saldo("John")); 
+console.log(obtener_saldo("Juan")); 
+console.log(obtener_saldo("John") + " cache"); 
+console.log(obtener_saldo("Jim")); 
+console.log(obtener_saldo("Jack"));
 
+
+
+//var z = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
+//var r = Math.ceil(Math.random() * 10);
 //  do you find any problem with this code? if so, what is the possible solution to prevent it?
 
-// YOUR
-// CODE
-// HERE
+const zeta: number[] = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
+const erre: number = Math.floor(Math.random() * zeta.length);
+console.log(`El elemento aleatorio de la matriz zeta es: ${zeta[erre]}`);
+
+console.log("Fin de los resultados...")
```

Comment:
don't remove the existing code, just modify it to be able to see this error in typescript.
```suggestion

```

Reply:
Se regreso a su estado original, solo se modifico el metodo ceil por el floor.

Reply:
the original code still missing